File: src/vct/col/ast/AxiomaticDataType.java
package vct.col.ast;

import java.util.ArrayList;
import java.util.Arrays;

import vct.util.ClassName;

public class AxiomaticDataType extends ASTDeclaration {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  private ArrayList<Method> cons=new ArrayList<Method>();
  private ArrayList<Method> maps=new ArrayList<Method>();
  private ArrayList<Axiom> axioms=new ArrayList<Axiom>();
  private DeclarationStatement pars[];
  
  public AxiomaticDataType(String name,DeclarationStatement ... pars) {
    super(name);
    this.pars=Arrays.copyOf(pars,pars.length);
  }

  public DeclarationStatement[] getParameters(){
    return pars;
  }
  
  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 

  @Override
  public ClassName getDeclName() {
    return new ClassName(name);
  }

  public Iterable<Method> constructors() {
    return cons;
  }
  
  public Iterable<Method> mappings() {
    return maps;
  }
  
  public Iterable<Axiom> axioms() {
    return axioms;
  }

  public void add_map(Method m){
    m.setFlag(ASTFlags.STATIC,true);
    m.setParent(this);
    maps.add(m);
  }
  
  public void add_cons(Method c){
    c.setFlag(ASTFlags.STATIC,true);
    c.setParent(this);
    cons.add(c);
  }
  
  public void add_axiom(Axiom ax){
    ax.setParent(this);
    axioms.add(ax);
  }
}

File: vercors/src/main/java/vct/col/ast/stmt/decl/AxiomaticDataType.scala
package vct.col.ast.stmt.decl

import vct.col.ast.generic.ASTNode

import scala.collection.JavaConverters._
import vct.col.ast.util.ASTVisitor
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}
import vct.util.ClassName

import scala.collection.mutable.ArrayBuffer

/**
 * @author sccblom, whmoortwijn
 * @note We may refactor the three mutable buffers if we first refactor `CheckHistoryAlgebra`.
 */
case class AxiomaticDataType(override val name:String, val parameters:List[DeclarationStatement]) extends ASTDeclaration(name) with VisitorHelper {
  require(parameters != null, "The list of parameters is null")

  /** Constructs a new ADT from Java constructs */
  def this(name:String, pars:Array[DeclarationStatement]) = this(name, pars.toList)

  /** Provides a Java wrapper (as an `java.util.List`) for the list of parameters. */
  def parametersJava = parameters.asJava

  private val axioms = new ArrayBuffer[Axiom]()
  private val constructors = new ArrayBuffer[Method]()
  private val mappings = new ArrayBuffer[Method]()

  def axiomsJava = axioms.toIterable.asJava
  def constructorsJava = constructors.toIterable.asJava
  def mappingsJava = mappings.toIterable.asJava

  def add_map(m:Method) : Unit = {
    m.setFlag(ASTFlags.STATIC, true)
    m.setParent(this)
    mappings += m
  }

  def add_cons(m:Method) : Unit = {
    m.setFlag(ASTFlags.STATIC, true)
    m.setParent(this)
    constructors += m
  }

  def add_axiom(ax:Axiom) : Unit = {
    ax.setParent(this);
    axioms += ax
  }

  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def getDeclName = new ClassName(name)

  override def debugTreeChildrenFields(): Iterable[String] = Seq("parameters")
  override def debugTreePropertyFields(): Iterable[String] = Seq("name")
}
